### ISSUE TITLE
implement an async_scoped_session

### ISSUE THREAD

**Issue by gboor** [sqlalchemy/sqlalchemy#6583]
**Describe the bug**
When using AsyncSession in a session_maker and scoped_session around that, the .close() call on the AsyncSession object is not awaited.

**To Reproduce**

```py
from asyncio import current_task, run

from sqlalchemy.ext.asyncio import create_async_engine, AsyncSession
from sqlalchemy.orm import scoped_session, sessionmaker


engine = create_async_engine('sqlite+aiosqlite:///test.db')
async_session = scoped_session(sessionmaker(engine, expire_on_commit=False, class_=AsyncSession),
                               scopefunc=current_task)


async def main():
    async with async_session() as session:
        pass

    async_session.remove()


run(main())
```

**Error**
Provide the complete text of any errors received **including the complete
stack trace**.   If the message is a warning, run your program with the
``-Werror`` flag:   ``python -Werror myprogram.py``

```
D:\dev\projects\async-test\site-packages\sqlalchemy\orm\scoping.py:145: RuntimeWarning: coroutine 'AsyncSession.close' was never awaited
  self.registry().close()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

**Versions.**
 - OS: Windows 10
 - Python: 3.9.2
 - SQLAlchemy: 1.4.15
 - Database: -
 - DBAPI: aiosqlite 0.17.0

**Additional context**
Test is provided using aiosqlite, because it's simple, but this fails using asyncpg as well.

**Have a nice day!**

**Comment by CaselIT** [comment]
Hi,

I may be wrong there, but I don't think scoped_session is supposed to work with the async extension, since it uses thread locals, not context vars.
That said we could raise an error or warn (and document it) when when trying to use it with an async session to indicate it's not compatible

@zzzeek thoughts?

**Comment by gboor** [comment]
@CaselIT I know that it uses thread locals by default, but there is an option to override the scopefunc. In my example I set that to current_task, so the session is local to the running asyncio task and not the current thread.

**Comment by CaselIT** [comment]
Part of the issue is that with the async extension `close` and other similar methods need to be awaited so probably some changes are needed to the interface to handle them

**Comment by gboor** [comment]
That was my thought as well. It might not even fit within the same interface/implementation. I worked around this now by implementing my own scoped_session using the ScopedRegistry. It's not super difficult, but given that you can pass AsyncSession to sessionmaker, it seems that this should work - or have a working alternative (like async_scoped_session or w/e).

**Comment by zzzeek** [comment]
there would need to be a whole new asyncscopedsession IIUC.

**Comment by zzzeek** [comment]
also "scopefunc" would be required for this object, no thread local

**Comment by jason3gb** [comment]
Hi all, I think the issue is mostly related to the remove method on scoped_session, which does not await the AsyncSession's close method in this case. @zzzeek I have a question that If we're going to implement a new async_scoped_session, should we just copy the existing methods in scoped_session and do some async modifications accordingly? I'm considering that might introduce some duplicated code. Also, @gboor are you working on this issue and submitting a PR? If not, may I try to work on this?

**Comment by CaselIT** [comment]
@zzzeek I think a similar issue (the close not awaited I mean) may present itself with the `close_all_sessions` function
but probably another issue it's best to track it

**Comment by gboor** [comment]
@jason3gb I just did a quick and dirty implementation of the same logic in my program. I am not working on a PR right now, feel free to take it if you have the time!

**Comment by jason3gb** [comment]
@CaselIT @zzzeek  I want to try implement this, is there a way to assign myself to this issue? like a "/assign" command?

**Comment by CaselIT** [comment]
@jason3gb not sure, I can manually assign you from the ui. I'm not sure if you also can or you need write access to the repo to do so.

**Comment by jason3gb** [comment]
> @jason3gb not sure, I can manually assign you from the ui. I'm not sure if you also can or you need write access to the repo to do so.

Thanks, I have no permission to assign myself. I'll start working on this issue.

**Comment by CaselIT** [comment]
Great, thanks!

**Comment by sqla-tester** [comment]
**jason3gb** has proposed a fix for this issue in the **master** branch:

**Fixes: #6583** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2870
